File: soske/comms/set.go
package comms

import (
	"fmt"
	"strings"

	"github.com/jmoiron/sqlx"
)

// SetCommand sets the value of a new or existing name.
func SetCommand(db *sqlx.DB, args []string) ([]string, error) {
	if len(args) < 2 {
		return nil, fmt.Errorf("not enough arguments")
	}

	name := strings.TrimSpace(args[0])
	body := strings.Join(args[1:], "\n")
	code := "insert or ignore into Keys (name) values (?)"
	if _, err := db.Exec(code, name); err != nil {
		return nil, err
	}

	code = "insert into Vals (key, body) values (?, ?)"
	if _, err := db.Exec(code, name, body); err != nil {
		return nil, err
	}

	return nil, nil
}

File: soske/comms/set_.py
"""
Command definition for 'set'.
"""

import sqlite3
import time

import click
from soske.comms.app import soske
from soske.tools import data


@soske.command(name="set", short_help="Set a key.")
@click.argument("key")
@click.argument("values", nargs=-1)
@click.pass_obj
def set_(conn: sqlite3.Connection, key: str, values: str):
    """
    Set the value of a new or existing key.
    """

    key = key.lower()
    val = "\n".join(values)
    now = time.time()
    row = data.row(conn, "select * from GoodVals where key=?", key)

    if not row or row["body"] != val:
        data.exec(conn, "insert or ignore into Keys values (?, false, ?)", key, now)
        data.exec(conn, "insert into Vals values (?, ?, ?)", key, val, now)
